src/pkg/systemartifact/manager.go

```diff
@@ -0,0 +1,235 @@
+
+	// Read a system artifact described by repository name and digest.
+	// The reader is responsible for closing the IO stream after the read completes.
+	Read(ctx context.Context, vendor string, repository string, digest string) (io.ReadCloser, error)
+
+	// Delete deletes a system artifact identified by a repository name and digest.
+	// Also deletes the tracking record from the underlying table.
+	Delete(ctx context.Context, vendor string, repository string, digest string) error
+
+	// Exists checks for the existence of a system artifact identified by repository and digest.
+	// A system artifact is considered as in existence if both the following conditions are true:
+	// 1. There is a system artifact tracking record within the Harbor DB
+	// 2. There is a BLOB corresponding to the repository name and digest obtained from system artifact record.
+	Exists(ctx context.Context, vendor string, repository string, digest string) (bool, error)
+
+	// GetStorageSize returns the total disk space used by the system artifacts stored in the registry.
+	GetStorageSize(ctx context.Context) (int64, error)
+
+	// RegisterCleanupCriteria a clean-up criteria for a specific vendor and artifact type combination.
+	RegisterCleanupCriteria(vendor string, artifactType string, criteria CleanupCriteria)
+
+	// GetCleanupCriteria returns a clean-up criteria for a specific vendor and artifact type combination.
+	// if no clean-up criteria is found then the default clean-up criteria is returned
+	GetCleanupCriteria(vendor string, artifactType string) CleanupCriteria
+
+	// Cleanup cleans up the system artifacts (tracking records as well as blobs) based on the
+	// artifact records selected by the CleanupCriteria registered for each vendor type.
+	// Returns the total number of records deleted, the reclaimed size and any error (if encountered)
+	Cleanup(ctx context.Context) (int64, int64, error)
+}
+
+type systemArtifactManager struct {
+	regCli                  registry.Client
+	dao                     dao.DAO
+	defaultCleanupCriterion CleanupCriteria
+	cleanupCriteria         map[string]CleanupCriteria
+	lock                    sync.Mutex
+}
+
+func NewManager() Manager {
+	sysArtifactMgr := &systemArtifactManager{
+		regCli:                  registry.Cli,
+		dao:                     dao.NewSystemArtifactDao(),
+		defaultCleanupCriterion: DefaultCleanupCriteria,
+		cleanupCriteria:         make(map[string]CleanupCriteria),
+	}
+	return sysArtifactMgr
+}
+
+func (mgr *systemArtifactManager) Create(ctx context.Context, artifactRecord *model.SystemArtifact, reader io.Reader) (int64, error) {
+
+	id, err := mgr.dao.Create(ctx, artifactRecord)
+
+	if err != nil {
+		return int64(0), err
+	}
+
+	repoName := mgr.getRepositoryName(artifactRecord.Vendor, artifactRecord.Repository)
+	err = mgr.regCli.PushBlob(repoName, artifactRecord.Digest, artifactRecord.Size, reader)
+
+	if err != nil {
```

Comment:
we can do the storage operation firstly, then DB. By doing this, no need to call the delete here.

Reply:
The reason I had to adopt this sequence is to ensure that there are no stray blobs within the repository that do not have a corresponding tracking record - can happen if we were to encounter a DB error when inserting the record. 
This makes the job of clean-up simpler as we only have to delete records that do not have a blob really stored on the registry filesystem.
To the best of my knowledge, there is no job/background processing right now that would automatically clean up aging blobs from the disk.
Also by calling Delete immediately following a failure to write the BLOB, we are able to ensure that our system space calculations account only for the BLOBS that are **actually** present on the disk as opposed to orphaned stray records in the DB.

Reply:
so, then we can use orm.WithTransaction(). When failure happens, the creation will be rolled back.

Reply:
Done. I have now moved the artifact record creation and the corresponding write to the blob store as a closure to the `WithTransaction` method.